### ISSUE TITLE
Eliminate large files from git history

### ISSUE THREAD

**Issue by r-barnes** [kellykochanski/rescal-snow#5]
The repo contains a number of large files that you likely wanted to ignore - the largest are listed below. This collectively means that the repo is a 100MB download.
```
41e6f427c11b  7.7MiB analysis/output_files/ALT_DATA2_OUT/fft/fft_results.gif
a8267f9be190  7.9MiB analysis/output_files/results_1/xcor/cross-correlations.txt
e271bcab6381   11MiB analysis/output_files/results_1/fft/fft_results.gif
669261e09a05   21MiB analysis/output_data/ALT_DATA1_OUT/xcor/cross-correlations.txt
36cbe3d82cf2   36MiB scripts/core.45511
4ac01836f00a   36MiB scripts/core.53132
9c2bb6f1759f   36MiB scripts/core.171982
a6cecc16b57b   57MiB analysis/output_data/ALT_DATA1_OUT/fft/fft_analysis_animation.gif
6def6506d3f7   66MiB scripts/GENESIS.log
```
these can be removed using the [BFG repo cleaner](https://rtyley.github.io/bfg-repo-cleaner/) using the following commands:
```
git clone --mirror https://github.com/kellykochanski/rescal-snow.git
java -jar ~/Downloads/bfg-1.12.13.jar --delete-folders 'output_files'  rescal-snow.git
java -jar ~/Downloads/bfg-1.12.13.jar --delete-folders 'output_data'  rescal-snow.git
java -jar ~/Downloads/bfg-1.12.13.jar --delete-files 'core.*'  rescal-snow.git
java -jar ~/Downloads/bfg-1.12.13.jar --delete-files 'GENESIS.log'  rescal-snow.git
java -jar ~/Downloads/bfg-1.12.13.jar --delete-files '*.o'  rescal-snow.git
java -jar ~/Downloads/bfg-1.12.13.jar --delete-files '*.py~'  rescal-snow.git
#Perhaps the `scripts/DUN.csp` file is also a temporary? It takes up 10MB.
```
after which you should check to make sure things look alright and then
```
cd rescal-snow.git
git reflog expire --expire=now --all && git gc --prune=now --aggressive
```
The upside is that this reduces the repo size to either 11MB (with `DUN.csp`) or (1MB without `DUN.csp`), which saves bandwidth and space for users.

**Comment by kellykochanski** [comment]
Q: several of those large files are graphics useful for examples and
documentation.

How can I leave them in, say, the online readme without requiring readers
to download them?

**Comment by r-barnes** [comment]
They must be in the repo to appear in the readme, unless you host them elsewhere.

However, none of the files I've suggested purging (I don't think) are currently used by the repo. These are (I think) all large files that were mistakenly committed in the past. The files you show on the readme are stored in `example_images` and take only 3.2MB.

**Comment by r-barnes** [comment]
@kellykochanski: I thought we were fixing this prior to JOSS?

**Comment by kellykochanski** [comment]
@rbarnes I haven't had time to get to it, and don't want to rush into messing with the git history.

**Comment by r-barnes** [comment]
Okay. Can we chat about it prior to JOSS acceptance?

**Comment by kellykochanski** [comment]
@r-barnes I used bfg as you suggested, and the repo is now 14MB (including the removal of DUN.csp - I think some additional docs with figures have been added since you opened this).

**Comment by r-barnes** [comment]
Doing this before merging outstanding PRs could make doing so impossible or
difficult...

**Comment by kellykochanski** [comment]
@bfg warned me... Any issue with just repeating the bfg calls after accepting them?

**Comment by zbeekman** [comment]
I just went through a similar process with another repository, although the issue was more related to pruning & relocating sensitive information prior to open-sourcing a software package. I discovered that GitHub has write protected refs for PRs. This means that you cannot prune data from these by default.

However, I think I have special settings in my git config to fetch these PR refs that most users do not have, so this may not be a real issue (at least not if you're only concerned about repo file size; it certainly is when you're removing sensitive info).

If it turns out that the PR refs keep the repository size bloated, then, the only solutions are either:

1) Contacting GitHub support and asking them to delete the old PR refs (I'm not sure if they can/will do this for you)
2) Deleting and recreating the repository.

Hopefully you won't need to do either and the PR refs won't much this up for you.

**Comment by r-barnes** [comment]
@zbeekman: Cool idea! So that cleans the while repo and associated PRs all
at once?

**Comment by zbeekman** [comment]
@r-barnes 

> Cool idea! So that cleans the while repo and associated PRs all at once?

Not 100% sure what you're talking about here. If it's my point 2. "Deleteing and recreating the repository" then I need to explain a little bit further:

What I really mean, is:

1. Move/rename the original repository (or at least keep a local backup clone as a copy, in addition to the one you plan to run BFG on, and then delete the original)
2. See if you have any PR refs in your local bare/mirrored repository with `git show-ref`
3. Use `git update-ref -d refs/.../... # protected github PR refs
4. Run BFG to eliminate bloat
5. Run the `git reflog` and `git gc` commands recommended by BFG
6. Create a new empty repo
7. `git push --mirror` or whatever BFG recommends to the new repo

__I would not recommend this__, unless the repo size stays large after a normal pass with BFG. Even then, it's much easier to contact GitHub support and ask if they can delete the old protected PR refs.

I had to go through this procedure because I realized that upon open sourcing a repository, you could still access old PR refs which included the sensitive information that cannot be made public. If you do not need to do it, then please don't.

Also, if you haven't run BFG yet to prune history, you may want to do it either before the final submission or not at all; I'm not sure if it will mess with JOSS' machinery, DOI process, etc. and it will certainly affect tagging.

**Comment by kellykochanski** [comment]
@zbeekman I ran bfg on the repository, though the changes were rejected from the then-open PR on kk/JOSS-fixes. Downloading rescal-snow is now down to 14MB from ~100MB.

I expect to have all open PRs closed at the time of JOSS acceptance, and will re-run bfg then - I can do this after finishing the corrections in your review, and merging the kk/JOSS-fixes branch, but before formal JOSS acceptance.

I hope bfg will work smoothly if all PRs are closed... Let me know if you think that it won't.

**Comment by zbeekman** [comment]
@kellykochanski: Yes it should work fine. IMO, you have images and stuff for the tutorials, and 14MB is probably how much space everything you want to keep takes up. But at the end of the day, I wouldn't bother with any steps that are more complicated than what you are doing. If you get complaints about rejected refs when you try to push due to PR refs, you can just delete them locally then try pushing again. (They will persist on the GitHub side, but I suspect this is fine and most people don't fetch them.)

**Comment by r-barnes** [comment]
@zbeekman: The issue is that the repo's history contains ~86MB worth of
large temporary and output files which we accidentally committed and later
removed.

**Comment by zbeekman** [comment]
@r-barnes: I'll pipe down and let you guys figure out what you want to do. My point was that *it sounds like Kelly had success with BFG* and *got things down to 14MB*, and __deleting__ the *entire* github repository and re-creating it is (hopefully) beyond the scope of what you want/need to accomplish. At any rate, sorry for the confusion and feel free to ignore my previous comments.

If you run into troubles pushing back up to github after running BFG, let me know, it might be the PR refs issue, and I may know the solution. Either way I'd happily take a look.

**Comment by r-barnes** [comment]
@zbeekman: No worries, thanks for your help.
